lib/java/persistance/src/main/java/org/enso/persist/PerGenerator.java

```diff
@@ -117,6 +182,15 @@ private static final class ReferenceOutput extends DataOutputStream
 
     @Override
     public <T> void writeInline(Class<T> clazz, T t) throws IOException {
+      if (Persistance.Reference.class == clazz) {
+        if (t instanceof Persistance.Reference<?> ref) {
+          var id = this.generator.registerReference(ref);
+          writeInt(id);
+          return;
+        } else {
+          throw new ClassCastException("Expecting Refernece");
+        }
```

Comment:
```suggestion
        Persistance.Reference<?> ref = (Persistance.Reference<?>) t;
        var id = this.generator.registerReference(ref);
        writeInt(id);
        return;
```

nitpick: I think we could rewrite this to a direct cast without the check - the behaviour is the same - if we the type is incorrect, a `ClassCastException` will be thrown - with a more detailed message (the default message not only tells us the expected type but also the actual type which may help recognize the problem source).